bug: delete existing .git directory if push-to-git and overwrite are set

ishitasequeira opened the pull request:
Resolves issue https://github.com/redhat-developer/kam/issues/212

**What type of PR is this?**
/kind bug


**Why we need it**:

The bootstrap command  
```
kam bootstrap --service-repo-url https://github.com/ishitasequeira/taxi --gitops-repo-url https://github.com/ishitasequeira/gitops.git --image-repo quay.io/isequeir/taxi --dockercfgjson isequeir-robot-auth.json --git-host-access-token ******* --output resources --overwrite --push-to-git=true
``` 
when applied to the existing output directory, the folder gets updated but the remote origin does not get updated according to the new repo and the bootstrap command fails with below error.

**What does this PR do**:
In this PR, we fix the above issue. While executing the bootstrap command, we check if the overwrite flag is set  or not set, `--push-to-git=true` and if the `.git` directory exists in the OutputPath. for `push-to-git=true`, if the overwrite flag is not set and the `.git` directory is present, we return an error and If the overwrite flag is set, we remove the existing `.git` directory before performing `git init` in the OutputPath.

**Have you updated the necessary documentation?**

* [ ] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #212 

**How to test changes / Special notes to the reviewer**:

**Pre-Steps:**
1. Execute the initial bootstrap command for the first time
```
kam bootstrap --service-repo-url https://github.com/ishitasequeira/taxi.git --gitops-repo-url https://github.com/<username>/gitops.git  --image-repo quay.io/isequeir/image-repo --dockercfgjson ~/Downloads/<username>-robot-auth.json --git-host-access-token ****  --output ~/go/src/github.com/test/gitops --push-to-git=true
```

2. The above command will succeed successfully as it is executed for the first time.
3. Delete the `pipelines.yaml` and the secrets directory from the OutputPath. 

**Scenario 1:** The .git directory exists in the OutputPath and the overwrite flag is not set with `push-to-git=true`:

4. Delete the repository from GitHub created in above command
5. Execute the same command as above: (without the overwrite flag)
```
kam bootstrap --service-repo-url https://github.com/ishitasequeira/taxi.git --gitops-repo-url https://github.com/<username>/gitops.git  --image-repo quay.io/isequeir/image-repo --dockercfgjson ~/Downloads/<username>-robot-auth.json --git-host-access-token ****  --output ~/go/src/github.com/test/gitops --push-to-git=true
```

The above command will give below error and the bootstrap command will fail as the .git directory is present in the OutputPath. Also, as the above command failed, it does not create the repository.

```
 ✗  .git directory in output path already exists. If you want to replace your existing files, please rerun with --overwrite
```


**Scenario 2:** The .git directory exists in the OutputPath and the overwrite flag is set and `push-to-git=true`:

 6. Execute the below command 
```
kam bootstrap --service-repo-url https://github.com/ishitasequeira/taxi.git --gitops-repo-url https://github.com/<username>/gitops.git  --image-repo quay.io/isequeir/image-repo --dockercfgjson ~/Downloads/<username>-robot-auth.json --git-host-access-token ****  --output ~/go/src/github.com/test/gitops  --overwrite --push-to-git=true
```
The above command should now complete successfully .

 7. To confirm the remote was correctly set, execute ` git remote -v` in the OutputPath
```
$ git remote -v
origin  git@github.com:ishitasequeira/gitops.git (fetch)
origin  git@github.com:ishitasequeira/gitops.git (push)
```

**Scenario 3:** The .git directory exists in the OutputPath and `push-to-git=false`

8. Execute the below command
```
./kam bootstrap --service-repo-url https://github.com/ishitasequeira/taxi.git --gitops-repo-url https://github.com/ishitasequeira/gitops.git  --image-repo quay.io/isequeir/image-repo --dockercfgjson ~/Downloads/isequeir-robot-auth.json --git-host-access-token 5d51786484454160e5418b4d35e4ae257d6487f3  --output ~/go/src/github.com/openshift/redhat-developer/test
```
The above command succeeds even if the OutputPath contains `.git` directory as `push-to-git=false`.



***Interactive mode***

The interactive command fails with below error message if we decide to push to git and the OutputPath contains the `.git` directory.
```
$ ./kam bootstrap
  
Checking dependencies

 ✓  Checking if Sealed Secrets is installed with the default configuration [114ms]
 ✓  Checking if ArgoCD is installed with the default configuration [227ms]
 ✓  Checking if OpenShift Pipelines Operator is installed with the default configuration [263ms]
  
Starting interactive prompt

? Do you want to accept all default values and be prompted only for the minimum required options? no
? You are able to seal secrets. Select Sealed to continue or Unsealed to generate unsealed secrets, which is not recommended. Sealed
? Provide the URL for your GitOps repository https://github.com/ishitasequeira/gitops.git
? Select type of image registry Openshift Internal registry
? Image registry of the form <project>/<app> which is used to push newly built images. test/app
? Provide a secret (minimum 16 characters) that we can use to authenticate incoming hooks from your Git hosting service for repository: https://github.com/ishitasequeira/gitops.git. (if not provided, it will be auto-generated) [? for help] 
? Provide the URL for your Service repository e.g. https://github.com/organisation/service.git https://github.com/ishitasequeira/taxi.git
? Provide a secret (minimum 16 characters) that we can use to authenticate incoming hooks from your Git hosting service for repository: https://github.com/ishitasequeira/taxi.git. (if not provided, it will be auto-generated) [? for help] 
? Please provide a token used to authenticate requests to "https://github.com/ishitasequeira/taxi.git" [? for help] ****************************************
? Do you wish to securely store the git-host-access-token in the keyring on your local machine? no
? Do you want to create and push the resources to your gitops repository? yes
? Add a prefix to the environment names(dev, stage, cicd etc.) to distinguish and identify individual environments? 
? Provide a path to write GitOps resources? /Users/ishitasequeira/go/src/github.com/openshift/redhat-developer/test
 ✗  the .git folder in output path /Users/ishitasequeira/go/src/github.com/openshift/redhat-developer/test already exists. Delete or rename the .git folder and try again
```

amitkrout commented on pkg/pipelines/utils_test.go:
```diff
@@ -113,6 +118,64 @@ func TestPushRepository(t *testing.T) {
+	e := newMockExecutor(outputs...)
+	testErr := errors.New("test error")
+	e.errors.push(testErr)
+	e.errors.push(nil)
+	e.errors.push(nil)
+	e.errors.push(nil)
+	e.errors.push(nil)
```
Duplicate lines, no ?

ishitasequeira replied:
Need to push the error messages as nil to the stack so that it gives required error only for the targeted check..

amitkrout commented:
Verified locally. Looks good. 
/approve

@chetan-rns Can you please review it implementation point of view

shubhamagarwal19 commented on pkg/pipelines/utils.go:
```diff
@@ -86,8 +86,14 @@ func pushRepository(o *BootstrapOptions, remote string, e executor) error {
-	if out, err := e.execute(o.OutputPath, "git", "remote", "add", "origin", remote); err != nil {
-		return fmt.Errorf("failed add remote 'origin' %q to repository in %q %q: %s", remote, o.OutputPath, string(out), err)
+	if _, err := e.execute(o.OutputPath, "git", "remote", "show", "origin"); err != nil {
+		if out, err := e.execute(o.OutputPath, "git", "remote", "add", "origin", remote); err != nil {
```
we can probably just have an error check here? 
if `err contains "remote origin already exists"` we can call `git remote set-url`, if the error is something else just return the error.
What do you think?

chetan-rns commented:
@amitkrout I think this PR is solving #212 rather than #207. WDYT?

shubhamagarwal19 commented:
> @amitkrout I think this PR is solving #212 rather than #207. WDYT?

@chetan-rns, yes !! it definitely fixes #212, I am not really sure what exactly is the error with issue #207!

wtam2018 commented:
Please update this PR to correctly point to which issue it is fixing.   Please update the description as well.   BTW, I don't think #212 is a bug.

ishitasequeira commented:
> Please update this PR to correctly point to which issue it is fixing. Please update the description as well. BTW, I don't think #212 is a bug.
@wtam2018  I have updated the PR to https://github.com/redhat-developer/kam/issues/212. https://github.com/redhat-developer/kam/issues/212 seems to be a bug since the overwrite flag replaces the existing bootstrap directory to the new one. When the output directory gets updated with the new one, the remote alias origin does not get updated. As we are using remote add, there is a remote url origin already present and it needs to be updated with the new url.

shubhamagarwal19 reviewed (approved):
Thanks @ishitasequeira. 
/LGTM

wtam2018 commented:
Please update "how to test"

ishitasequeira commented:
> Please update "how to test"

I have updated

wtam2018 commented:
We don't know what else is in the .git.  It is better off to check if .git exist.  If it does exist, we should fail the command with the appropriate error.   If there is an overwrite flag, we can remove the .git folder.

wtam2018 commented:
/hold

wtam2018 commented:
When push-to-git was implemented, one of the assumptions was that the git repo had not been initialized.   Now, you are running a sequence of git commands with an initialized repo in an unknown state.

shubhamagarwal19 commented:
> When push-to-git was implemented, one of the assumptions was that the git repo had not been initialized. Now, you are running a sequence of git commands with an initialized repo in an unknown state.

@wtam2018, this would only happen when the user explicitly provides `--overwrite` flag while bootsrapping. 
Right now, when the overwrite flag is provided, it updates the local output directory but doesn't push it to the git which is not consistent. 
So IMO, if the user is OK to overwrite the local output directory, it should be OK to update the git repo as well.

shubhamagarwal19 commented on pkg/pipelines/utils.go:
```diff
@@ -87,7 +87,13 @@ func pushRepository(o *BootstrapOptions, remote string, e executor) error {
 	}
 	if out, err := e.execute(o.OutputPath, "git", "remote", "add", "origin", remote); err != nil {
-		return fmt.Errorf("failed add remote 'origin' %q to repository in %q %q: %s", remote, o.OutputPath, string(out), err)
+		if strings.Contains(string(out), "remote origin already exists") {
```
can you check for `overwrite` flag here as well?

shubhamagarwal19 commented:
> We don't know what else is in the .git. It is better off to check if .git exist. If it does exist, we should fail the command with the appropriate error. If there is an overwrite flag, we can remove the .git folder.

aaahhh i missed this comment. 
Yes! it actually does make more sense to check for `.git` directory and `overwirte` flag. If `overwrite=true` delete the `.git` folder and reinitialize it.

amitkrout commented:
/refresh

amitkrout commented:
/retest

ishitasequeira commented:
> When push-to-git was implemented, one of the assumptions was that the git repo had not been initialized. Now, you are running a sequence of git commands with an initialized repo in an unknown state.

@wtam2018 I have addressed the comments

wtam2018 commented:
@ishitasequeira  I don't think "How to test changes " is correct.  You need to delete the gitops remote after step 4.   "Both commands should execute successfully" does not seem to make sense,

wtam2018 commented on pkg/pipelines/utils.go:
```diff
@@ -74,6 +77,12 @@ func BootstrapRepository(o *BootstrapOptions, f clientFactory, e executor) error
 }
 
 func pushRepository(o *BootstrapOptions, remote string, e executor) error {
+	exists, _ := ioutils.IsExisting(ioutils.NewFilesystem(), filepath.Join(o.OutputPath, ".git"))
```
1.  We should only do so if --overwrite is true.
2.  If --overwrite is false and --push-to-git is true and .git exists, we should return a error message.   If possible, an error is return before we start performing "push to git" (i.e. we should not true to create a new git repo if --overwrite is false and .git exists)

shubhamagarwal19 replied:
@ishitasequeira, are you assuming that if `--overwrite=false` and `output` directory exists then it would error out even before it tries to update the contents of the `output` directory? 
If that is true, i think its OK to not have the `--overwrite` check here!

ishitasequeira replied:
@wtam2018 As discussed today, I have addressed the comments. I have updated the PR description as well.
@shubhamagarwal19 I have updated the workflow to check for overwrite flag and `.git` directory and return an error before creating the github repository if the directory exists and overwrite flag is false.

wtam2018 commented on pkg/pipelines/bootstrap_test.go:
```diff
@@ -374,6 +376,37 @@ func TestOverwriteFlag(t *testing.T) {
+		ServiceWebhookSecret: "456",
+		OutputPath:           "/tmp",
+	}
+	err := os.MkdirAll(filepath.Join(params.OutputPath, ".git"), 0755)
```
```suggestion
	err := fakeFs.MkdirAll(filepath.Join(params.OutputPath, ".git"), 0755)
```

wtam2018 commented on pkg/pipelines/bootstrap.go:
```diff
@@ -471,6 +471,11 @@ func checkPipelinesFileExists(appFs afero.Fs, outputPath string, overWrite bool)
+	}
 	return nil
 }
 
```
We should clean this up.
```
// Checks whether the pipelines.yaml is present in the output path specified.
func checkPipelinesFileExists(appFs afero.Fs, outputPath string, overWrite bool) error {

	if overWrite {
		return nil
	}

	err := errorIfFileExists(appFs, outputPath, pipelinesFile, ".git")
	if err != nil {
		return err
	}

	secretsFolderExists, _ := ioutils.IsExisting(appFs, filepath.Join(outputPath, "..", "secrets"))
	if secretsFolderExists {
		return fmt.Errorf("the secrets folder located as a sibling of the output folder %s already exists. Rerun with --overwrite", outputPath)
	}

	return nil
}

func errorIfFileExists(appFs afero.Fs, outputPath string, files ...string) error {
	for _, file := range files {
		exists, _ := ioutils.IsExisting(appFs, filepath.Join(outputPath, file))
		if exists {
			return fmt.Errorf("%s in output path already exists. If you want to replace your existing files, please rerun with --overwrite", file)
		}
	}
	return nil
}
```

wtam2018 commented on pkg/pipelines/utils.go:
```diff
@@ -74,6 +77,11 @@ func BootstrapRepository(o *BootstrapOptions, f clientFactory, e executor) error
 
 func pushRepository(o *BootstrapOptions, remote string, e executor) error {
+	if exists, _ := ioutils.IsExisting(ioutils.NewFilesystem(), filepath.Join(o.OutputPath, ".git")); exists {
+		if err := os.RemoveAll(filepath.Join(o.OutputPath, ".git")); err != nil {
```
We should use afero.FS.

wtam2018 commented:
Please also test interactive mode.

wtam2018 commented on pkg/pipelines/bootstrap.go:
```diff
@@ -461,16 +462,37 @@ func defaultPipelines(r scm.Repository) *config.Pipelines {
+		return nil
+	}
+
+	if pushToGit {
+		if err := errorIfFileExists(appFs, outputPath, ".git"); err != nil {
+			return err
+		}
+	}
+
+	if err := errorIfFileExists(appFs, outputPath, pipelinesFile); err != nil {
+		return err
 	}
```
```suggestion
	if pushToGit {
		checkList = append(checkList, ".git")
	}

	if err := errorIfFileExists(appFs, outputPath, checkList...); err != nil {
		return err
	}
```

wtam2018 commented on pkg/cmd/bootstrap.go:
```diff
@@ -235,7 +234,12 @@ func initiateInteractiveMode(io *BootstrapParameters, client *utility.Client, cm
+		}
+	}
 	if !io.Overwrite {
 		exists := ui.VerifySecretsPath(io.OutputPath)
```
move "if io.PushToGit {}" to here.

ishitasequeira replied:
@wtam2018 Thanks for the review. Updated as you mentioned.

ishitasequeira commented:
/refresh

ishitasequeira commented:
/retest

wtam2018 commented on pkg/cmd/bootstrap.go:
```diff
@@ -235,12 +234,16 @@ func initiateInteractiveMode(io *BootstrapParameters, client *utility.Client, cm
 	io.OutputPath, io.Overwrite = ui.VerifyOutputPath(io.OutputPath, io.Overwrite, outputPathOverridden, promptForAll)
-
 	if !io.Overwrite {
 		exists := ui.VerifySecretsPath(io.OutputPath)
 		if exists {
 			return fmt.Errorf("the secrets folder located as a sibling of the output folder %s already exists. Delete or rename the secrets folder and try again", io.OutputPath)
 		}
```
```suggestion
		if exists := ui.VerifySecretsPath(io.OutputPath); exists {
			return fmt.Errorf("the secrets folder located as a sibling of the output folder %s already exists. Delete or rename the secrets folder and try again", io.OutputPath)
		}
```

wtam2018 commented on pkg/cmd/bootstrap.go:
```diff
@@ -235,12 +234,16 @@ func initiateInteractiveMode(io *BootstrapParameters, client *utility.Client, cm
 	io.OutputPath, io.Overwrite = ui.VerifyOutputPath(io.OutputPath, io.Overwrite, outputPathOverridden, promptForAll)
-
 	if !io.Overwrite {
 		exists := ui.VerifySecretsPath(io.OutputPath)
 		if exists {
 			return fmt.Errorf("the secrets folder located as a sibling of the output folder %s already exists. Delete or rename the secrets folder and try again", io.OutputPath)
 		}
+		if io.PushToGit {
+			if exists := ui.VerifyGitPath(io.OutputPath); exists {
+				return fmt.Errorf("the .git folder in output path %s already exists. Delete or rename the .git folder and try again", io.OutputPath)
+			}
+		}
```
```suggestion
		if ui.PathExists(filepath.Join(io.OutputPath, "..", "secrets")) {
			return fmt.Errorf("the secrets folder located as a sibling of the output folder %s already exists. Delete or rename the secrets folder and try again", io.OutputPath)
		}
		if io.PushToGit && ui.PathExists(filepath.Join(io.OutputPath, ".git")) {
			return fmt.Errorf("the .git folder in output path %s already exists. Delete or rename the .git folder and try again", io.OutputPath)
		}
		}
```

wtam2018 commented on pkg/cmd/ui/ui.go:
```diff
@@ -108,6 +108,12 @@ func VerifyOutputPath(originalPath string, overwrite, outputPathOverridden, prom
 	return outputPath, doOverwrite
 }
 
+func VerifyGitPath(outputPath string) bool {
+	exists, err := ioutils.IsExisting(ioutils.NewFilesystem(), filepath.Join(outputPath, ".git"))
+	handleError(err)
+	return exists
+}
```
```suggestion
func PathExists(path string) bool {
	exists, err := ioutils.IsExisting(ioutils.NewFilesystem(), path)
	handleError(err)
	return exists
}
```

wtam2018 commented on pkg/cmd/ui/ui.go:
```diff
@@ -108,6 +108,12 @@ func VerifyOutputPath(originalPath string, overwrite, outputPathOverridden, prom
+	handleError(err)
+	return exists
+}
+
```
Remove VerifySecretsPath()
change the ioutils func to use aferofs

ishitasequeira replied:
Will remove verifySecretsPath().
ioutils.NewFileSystem() returns an instance of aferofs. Are you suggesting to replace it?

wtam2018 commented on pkg/pipelines/bootstrap.go:
```diff
@@ -92,6 +92,7 @@ type BootstrapOptions struct {
 	PrivateRepoDriver        string               // Records the type of the GitOpsRepoURL driver if not a well-known host.
 	CommitStatusTracker      bool                 // If true, this is a "private repository", i.e. requires authentication to clone the repository.
 	Insecure                 bool                 // If true, use unencrypted, unsealed secrets. By default, sealed secrets are generated.
+	PushToGit                bool                 // records whether or not the repository should be pushed to git.
```
```suggestion
	PushToGit                bool                 // If true, gitops repository is pushed to remote git repository.
```

wtam2018 reviewed (commented):
/lgtm
/approve

wtam2018 commented:
/hold cancel
